tests/unit/test_endpoints_v2.py

```diff
@@ -0,0 +1,409 @@
+    assert endpoint_rule.partition(None) is None
+
+
+def test_boolean_equals_raises(endpoint_rule):
+    with pytest.raises(EndpointInputParametersError) as err:
+        endpoint_rule.boolean_equals(1, 2)
+        assert err.msg == 'Both arguments must be booleans'
+
+
+def test_substring_raises(endpoint_rule):
+    with pytest.raises(EndpointInputParametersError) as err:
+        endpoint_rule.substring(['h', 'e', 'l', 'l', 'o'], 1, 5, False)
+        assert err.msg == 'Input must be a string'
+
+
+@pytest.mark.parametrize(
+    "input_string,start,stop,reverse,expected_output",
+    [
+        ("hello", 0, 5, False, "hello"),
+        ("hello", 0, -1, False, None),
+        ("hello", -5, -1, False, "hell"),
+        ("hello", -5, 5, False, "hello"),
+        ("hello", 0, 10, False, None),
+        ("hello", 3, 4, False, "l"),
+    ],
+)
+def test_substring(
+    endpoint_rule, input_string, start, stop, reverse, expected_output
+):
+    assert (
+        endpoint_rule.substring(input_string, start, stop, reverse)
+        == expected_output
+    )
+
+
+@pytest.mark.parametrize(
+    "input_url,expected_url",
+    [
+        (
+            {'fn': 'getAttr', 'argv': [{'ref': 'testUrls'}, 'one[0]']},
+            'https://us-west-2.myGreatService.amazonaws.com',
+        ),
+        ({'ref': 'testUrl'}, 'https://www.amazon.com'),
+    ],
+)
+def test_resolve_url(endpoint_rule, input_url, expected_url):
+    endpoint_rule.endpoint['url'] = input_url
+    assert endpoint_rule.resolve_url() == expected_url
+
+
+def test_resolve_headers(endpoint_rule):
+    endpoint_rule.evaluate_conditions()
+    headers = endpoint_rule.resolve_headers()
+    assert headers == {
+        'x-amz-region-set': [
+            'us-west-2',
+            'us-west-2',
+            'us-east-2',
+        ],
+    }
+
+
+@pytest.mark.parametrize(
+    "error_rule,expected_error",
+    [
+        (
+            ErrorRule(
+                assignments={'errorMessage': 'This was bad'},
+                conditions=[],
+                provider=endpoint_provider,
+                error={'ref': 'errorMessage'},
+            ),
+            'This was bad',
+        ),
+        (
+            ErrorRule(
+                assignments={
+                    'errorMessages': {
+                        'important': ['Fix this', 'This was bad']
+                    }
+                },
+                conditions=[],
+                provider=endpoint_provider,
+                error={
+                    'fn': 'getAttr',
+                    'argv': [{'ref': 'errorMessages'}, 'important[0]'],
+                },
+            ),
+            'Fix this',
+        ),
+    ],
+)
+def test_resolve_error(error_rule, expected_error):
+    assert error_rule.resolve_error() == expected_error
+
+
+def test_validate_input():
+    param_spec = ParameterDefinition(
+        name='test',
+        type='String',
+    )
+    correct_type = getattr(param_spec.ParameterType, param_spec.type)
```

Comment:
Let's be concrete here instead of using the same logic as the function. We know it needs to be `str` so let's make sure to assert that. Otherwise, if there's a flaw in the ParameterType, this test won't catch it.

Reply:
yeah that makes sense I see how this is a bit circular logic